File: iosApp/iosApp/Modules/ImprovementProposals/ImprovementProposalsViewModel.swift
// Created by web3d3v on 30/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import UIKit

enum ImprovementProposalsViewModel {
    case loading
    case loaded(sections: [Section], selectedSectionType: Section.`Type`)
    case error(error: AppsViewModel.Error)
}

extension ImprovementProposalsViewModel {
    
    struct Section {
        let title: String
        let description: String
        let type: `Type`
        let items: [Item]
        let footer: Footer?
        
        enum `Type` {
            case all
            case infrastructure
            case integrations
            case features
            
            var stringValue: String {
                switch self {
                case .all:
                    return Localized("proposals.segmentedControl.all")
                case .infrastructure:
                    return Localized("proposals.segmentedControl.infrastructure")
                case .integrations:
                    return Localized("proposals.segmentedControl.integrations")
                case .features:
                    return Localized("proposals.segmentedControl.features")
                }
            }
            
            var descriptionValue: String {
                switch self {
                case .all:
                    return Localized("proposals.section.all.description")
                case .infrastructure:
                    return Localized("proposals.section.infrastructure.description")
                case .integrations:
                    return Localized("proposals.section.integrations.description")
                case .features:
                    return Localized("proposals.section.features.description")
                }
            }
        }
        
        struct Footer {
            let imageName: String
            let text: String
        }
    }

    struct Item {
        let id: String
        let title: String
        let subtitle: String
        let buttonTitle: String
    }
}

extension ImprovementProposalsViewModel {

    struct Error {
        let title: String
        let body: String
        let actions: [String]
    }
}

extension ImprovementProposalsViewModel {
    
    var title: String {
        Localized("proposals.title")
    }
    
    var sections: [ImprovementProposalsViewModel.Section] {
        
        switch self {
        case let .loaded(sections, _): return sections
        default: return []
        }
    }
    
    var selectedSectionType: ImprovementProposalsViewModel.Section.`Type` {
        switch self {
        case let .loaded(_, type): return type
        default: return .all
        }
    }
}

extension ImprovementProposalsViewModel.Section.`Type` {

    static func from(
        _ category: ImprovementProposal.Category
    ) -> ImprovementProposalsViewModel.Section.`Type` {
        switch category {
        case .infrastructure: return .infrastructure
        case .integration: return .integrations
        case .feature: return .features
        default: return .all
        }
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/improvementProposals/ImprovementProposalsViewModel.kt
package com.sonsofcrypto.web3walletcore.modules.improvementProposals

import com.sonsofcrypto.web3walletcore.common.viewModels.ErrorViewModel
import com.sonsofcrypto.web3walletcore.common.viewModels.SectionHeaderViewModel
import com.sonsofcrypto.web3walletcore.extensions.Localized
import com.sonsofcrypto.web3walletcore.services.improvementProposals.ImprovementProposal
import com.sonsofcrypto.web3walletcore.services.improvementProposals.ImprovementProposal.Category.*

sealed class ImprovementProposalsViewModel {
    /** Loading proposals */
    object Loading: ImprovementProposalsViewModel()
    /** Successfully fetched proposals */
    data class Loaded(
        val categories: List<Category>,
        val selectedIdx: Int
    ): ImprovementProposalsViewModel()
    /** Error occurred */
    data class Error(
        val error: ErrorViewModel
    ): ImprovementProposalsViewModel()

    data class Category(
        val title: String,
        val description: String,
        val items: List<Item>
    )

    data class Item(
        val id: String,
        val title: String,
        val subtitle: String,
    )

    fun selectedCategory(): Category? = when (this) {
        is Loaded -> if(categories.isNotEmpty()) categories[selectedIdx] else null
        else -> null
    }

    companion object
}

fun ImprovementProposal.Category.title(): String = when (this) {
    INTEGRATION -> Localized("proposals.infrastructure.title")
    INFRASTRUCTURE -> Localized("proposals.integration.title")
    FEATURE -> Localized("proposals.feature.title")
    UNKNOWN -> Localized("proposals.unknown.title")
} + " " + Localized("proposals")

fun ImprovementProposal.Category.description(): String = when (this) {
    INTEGRATION -> Localized("proposals.infrastructure.description")
    INFRASTRUCTURE -> Localized("proposals.integration.description")
    FEATURE -> Localized("proposals.feature.description")
    UNKNOWN -> ""
}
